crates/snapshot/src/snapshotter.rs

```diff
@@ -0,0 +1,172 @@
+//! Support for snapshotting.
+
+use crate::SnapshotterError;
+use reth_db::database::Database;
+use reth_primitives::{BlockNumber, ChainSpec, TxNumber};
+use reth_provider::{BlockReader, ProviderFactory};
+use std::sync::Arc;
+
+/// Result of [Snapshotter::run] execution.
+pub type SnapshotterResult = Result<(), SnapshotterError>;
+
+/// The snapshotter type itself with the result of [Snapshotter::run]
+pub type SnapshotterWithResult<DB> = (Snapshotter<DB>, SnapshotterResult);
+
+/// Snapshotting routine. Main snapshotting logic happens in [Snapshotter::run].
+pub struct Snapshotter<DB> {
+    provider_factory: ProviderFactory<DB>,
+    last_snapshots: Snapshots,
+    thresholds: SnapshotThresholds,
+}
+
+macro_rules! snapshots {
+    ($($name:ident: $ty:ty),* $(,)?) => {
+        #[derive(Debug, Clone, Copy)]
+        struct Snapshots {
+            $($name: $ty,)+
+        }
+
+        /// Thresholds for snapshotting per data part, measured in part-specific unit (usually
+        /// [BlockNumber] or [TxNumber]).
+        ///
+        /// When the specified number of entities accumulated in database since last snapshot for
+        /// data part, [Snapshotter::get_snapshot_request] will return [Option::Some].
+        #[derive(Debug, Clone, Copy)]
+        pub struct SnapshotThresholds {
+            $($name: u64,)+
+        }
+
+        /// Snapshot request with targets per data part, measured in part-specific unit (usually
+        /// [BlockNumber] or [TxNumber]).
+        #[derive(Debug, Clone, Copy)]
+        pub struct SnapshotRequest {
+            $($name: Option<$ty>,)+
+        }
+
+        impl SnapshotRequest {
+            /// Returns `true` if any of the data parts has targets, i.e. is [`Option::Some`].
+            pub fn any(&self) -> bool {
+                $(self.$name.is_some())||+
+            }
+        }
+    };
+}
+
+snapshots!(
+    headers: BlockNumber,
+    receipts: TxNumber,
+    transactions: TxNumber,
+);
```

Comment:
original idea is that they are all snapshotted at the same unit, in this case `BlockNumber`

Reply:
otherwise you might have a  transaction snapshot from block N -> N + 5 , and 2 block snapshots block N -> N + 2 and block N+3 -> block N+5.... and i'm speculating that orchestrating this in the `SnapshotProvider` becomes trickier.

Reply:
atomicity-wise, i think it's fine that we split them like this, as long as we have some kind of checkpoint (`finished headers. missing: receipts,transactions`).

however, i would make each snapshotter type atomic without any kind of checkpoint.

Reply:
agree, made all parts dependent and snapshotter by `BlockNumber` with optional `TxNumber` in `SnapshotTargets`.